device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/MultiplexingClient.java

```diff
@@ -0,0 +1,456 @@
+                for (DeviceClient currentClient : deviceClientList)
+                {
+                    String currentDeviceId = currentClient.getConfig().getDeviceId();
+                    if (currentDeviceId.equalsIgnoreCase(configToAdd.getDeviceId()))
+                    {
+                        deviceAlreadyRegistered = true;
+                        break;
+                    }
+                }
+
+                if (deviceAlreadyRegistered)
+                {
+                    log.debug("Device {} wasn't registered to the multiplexed connection because it is already registered.", configToAdd.getDeviceId());
+                    continue;
+                }
+
+                if (configToAdd.getAuthenticationType() != DeviceClientConfig.AuthType.SAS_TOKEN)
+                {
+                    throw new UnsupportedOperationException("Can only register to multiplex a device client that uses SAS token based authentication");
+                }
+
+                if (configToAdd.getProtocol() != this.protocol)
+                {
+                    throw new UnsupportedOperationException("A device client cannot be registered to a multiplexing client that specifies a different transport protocol.");
+                }
+
+                if (this.protocol == IotHubClientProtocol.AMQPS && this.deviceClientList.size() > MAX_MULTIPLEX_DEVICE_COUNT_AMQPS)
+                {
+                    throw new UnsupportedOperationException(String.format("Multiplexed connections over AMQPS only support up to %d devices", MAX_MULTIPLEX_DEVICE_COUNT_AMQPS));
+                }
+
+                // Typically client side validation is duplicate work, but IoT Hub doesn't give a good error message when closing the
+                // AMQPS_WS connection so this is the only way that users will know about this limit
+                if (this.protocol == IotHubClientProtocol.AMQPS_WS && this.deviceClientList.size() > MAX_MULTIPLEX_DEVICE_COUNT_AMQPS_WS)
+                {
+                    throw new UnsupportedOperationException(String.format("Multiplexed connections over AMQPS_WS only support up to %d devices", MAX_MULTIPLEX_DEVICE_COUNT_AMQPS_WS));
+                }
+
+                if (!this.hostName.equalsIgnoreCase(configToAdd.getIotHubHostname()))
+                {
+                    throw new UnsupportedOperationException("A device client cannot be registered to a multiplexing client that specifies a different host name.");
+                }
+
+                if (deviceClientToRegister.getDeviceIO() != null && deviceClientToRegister.getDeviceIO().isOpen())
+                {
+                    throw new IllegalStateException("Cannot register a device client to a multiplexed connection when it the device client was already opened.");
+                }
+
+                deviceClientToRegister.setAsMultiplexed();
+                deviceClientToRegister.setDeviceIO(this.deviceIO);
```

Comment:
I am a bit confused about how deviceIO works. This is what I thought, but now I think I might be misunderstanding it:
- deviceIO is the "interface" for transforming device client operations to the transport layer
- `this.deviceIO` is the deviceIO associated with the multiplexing client, which is then getting assigned to the device client instance.
In that case, how do we differentiate between the deviceIO for device1 vs device2?

Reply:
DeviceIO is a layer whose primary responsibility is managing the worker threads of the SDK. When multiplexing, all devices share the same deviceIO instance and the same worker threads. Within the deviceIO layer is a reference to the transport layer. The deviceIO layer more or less just forwards all telemetry/twin/method messages down to the transport layer. I hope that clears it up a bit

Reply:
> In that case, how do we differentiate between the deviceIO for device1 vs device2?

When multiplexing, they have the same deviceIO instance